pythran/analyses/range_values.py

```diff
@@ -0,0 +1,448 @@
+        Range(low=-inf, high=inf)
+        """
+        if isinstance(op, ast.Add):
+            min_, max_ = range1.low + range2.low, range1.high + range2.high
+        elif isinstance(op, ast.Sub):
+            min_, max_ = range1.low - range2.high, range1.high - range2.low
+        elif isinstance(op, ast.Mult):
+            res = [v1 * v2 for v1, v2 in itertools.product(range1, range2)]
+            min_, max_ = min(res), max(res)
+        elif isinstance(op, ast.Div):
+            if range2.low <= 0 and range2.high >= 0:
+                return RangeValues.UNKNOWN_RANGE
+            # This is valid because range2 doesn't change its sign
+            res = [v1 / v2 for v1, v2 in itertools.product(range1, range2)]
+            min_, max_ = min(res), max(res)
+        elif isinstance(op, ast.Mod):
+            min_, max_ = 0, min(range2.high, max(abs(range1.high),
+                                                 abs(range1.low)))
+        elif isinstance(op, ast.Pow):
+            res = [v1 ** v2 for v1, v2 in itertools.product(range1, range2)]
+            min_, max_ = math.ceil(min(res)), math.floor(max(res))
+        elif isinstance(op, ast.LShift):
+            min_inf = isinf(range1.low) or isinf(range2.low)
+            max_inf = isinf(range1.high) or isinf(range2.high)
+            min_ = -float("inf") if min_inf else (range1.low << range2.low)
+            max_ = float("inf") if max_inf else (range1.high << range2.high)
+        elif isinstance(op, ast.RShift):
+            if range1.low <= 0:
+                if isinf(range1.low):
+                    min_ = range1.low
+                else:
+                    min_ = range1.low >> range2.low
+            elif isinf(range2.high):
+                min_ = 0
+            else:
+                min_ = range1.low >> range2.high
+            if isinf(range1.high):
+                max_ = range1.high
+            elif isinf(range2.low):
+                max_ = 0
+            else:
+                max_ = range1.high >> range2.low
+        elif isinstance(op, (ast.BitOr, ast.BitXor, ast.BitAnd)):
+            return RangeValues.UNKNOWN_RANGE
+        else:
+            assert isinstance(op, ast.FloorDiv), "No possible other operation."
+            if range2.low <= 0 and range2.high >= 0:
+                return RangeValues.UNKNOWN_RANGE
+            res = [v1 // v2 for v1, v2 in itertools.product(range1, range2)]
+            min_, max_ = min(res), max(res)
+        return Range(min_, max_)
+
+    def visit_FunctionDef(self, node):
+        """ Set default range value for globals and attributs.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(a, b): pass")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=inf)
+        """
+        for global_name in self.globals:
+            self.result[global_name] = self.UNKNOWN_RANGE
```

Comment:
Yes because we need a value if we use it later.
I could have use a : `self.result.get(variable_name, UNKNOWN_RANGE)` in the visit_Name but this way, I have a better error report if I missed something.